File: spring-demoapp/src/main/java/com/example/springapp/service/MetadataService.java
package com.example.springapp.service;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

import com.example.springapp.aspect.LatencyLogger;
import com.example.springapp.data.Director;
import com.example.springapp.data.DirectorRepository;
import com.example.springapp.data.Movie;
import com.example.springapp.data.MovieRepository;
import com.example.springapp.types.CreateDirectorRequest;
import com.example.springapp.types.CreateMovieRequest;
import com.example.springapp.web.MetadataDto;

@Service
@Transactional
public class MetadataService {

	@Autowired
	private DirectorRepository directorRepo;

    @Autowired
    private MovieRepository movieRepo;

    @LatencyLogger
    public Director createDirector(final CreateDirectorRequest createDirectorRequest) {
        Director director = new Director();
        director.setName(createDirectorRequest.getName());
        director.setMovies(new ArrayList<>());
        director = directorRepo.save(director);
        return director;
    }

    @LatencyLogger
    public Movie createMovie(final CreateMovieRequest createMovieRequest) {
        Movie movie = new Movie();
        movie.setTitle(createMovieRequest.getTitle());
        Director director = directorRepo.findOneByName(createMovieRequest.getDirectorName());
        if(director == null && !StringUtils.isEmpty(createMovieRequest.getDirectorName())) {
            CreateDirectorRequest createDirectorRequest = new CreateDirectorRequest();
            createDirectorRequest.setName(createMovieRequest.getDirectorName());
            director = createDirector(createDirectorRequest);
        }
        if(director != null) {
            director.getMovies().add(movie);
            movie.setDirector(director);
        }
        movie.setYear(createMovieRequest.getYear());
        movie = movieRepo.save(movie);
        return movie;
    }

    public MetadataDto getMetadata(final Long id) {
        MetadataDto dto = null;
        Movie movie = movieRepo.findById(id).orElse(null);
        if (movie != null) {
            dto = new MetadataDto();
            dto.setMovieId(id);
            dto.setTitle(movie.getTitle());
            dto.setYear(movie.getYear());
            dto.setMovieVersion(movie.getVersion());
            dto.setMovieCreatedDate(movie.getCreatedDate().toString());

            Director director = movie.getDirector();
            if (director != null) {
                dto.setDirectorId(movie.getDirector().getId());
                dto.setDirectorName(movie.getDirector().getName());
                if (director.getMovies() != null) {
                    dto.setDirectorOtherMovies(
                            director.getMovies()
                                    .stream()
                                    .filter(m -> !m.getTitle().equalsIgnoreCase(movie.getTitle()))
                                    .map(m -> m.getTitle()).collect(Collectors.toList()));
                }
                dto.setDirectorVersion(director.getVersion());
                dto.setDirectorCreatedDate(director.getCreatedDate().toString());
            }
        }
        return dto;
    }

    public List<MetadataDto> getAllMetadata() {
    	return movieRepo.findAll()
    			.stream()
    			.map(movie -> getMetadata(movie.getId()))
    			.collect(Collectors.toList());
    }
}

File: spring-demoapp-kotlin/src/main/java/com/example/springapp/service/MetadataService.kt
package com.example.springapp.service

import com.example.springapp.data.Director
import com.example.springapp.data.DirectorRepository
import com.example.springapp.data.Movie
import com.example.springapp.data.MovieRepository
import com.example.springapp.types.CreateDirectorRequest
import com.example.springapp.types.CreateMovieRequest
import com.example.springapp.web.MetadataDto
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.StringUtils

@Service
@Transactional
class MetadataService(
        private val movieRepo: MovieRepository,
        private val directorRepo: DirectorRepository) {

    fun createMovie(createMovieRequest: CreateMovieRequest): Movie {
        var movie = Movie()
        movie.title = createMovieRequest.title
        var director: Director? = directorRepo.findOneByName(createMovieRequest.directorName)
        if (director == null && !StringUtils.isEmpty(createMovieRequest.directorName)) {
            val createDirectorRequest = CreateDirectorRequest()
            createDirectorRequest.name = createMovieRequest.directorName
            director = createDirector(createDirectorRequest)
            director.movies =  mutableListOf()
        }
        if (director != null) {
            director.movies?.add(movie)
            movie.director = director
        }
        movie.year = createMovieRequest.year
        movie = movieRepo.save(movie)
        return movie
    }

    fun createDirector(createDirectorRequest: CreateDirectorRequest): Director {
        var director = Director()
        director.name = createDirectorRequest.name
        director = directorRepo.save(director)
        return director
    }

    fun getMetadata(id: Long): MetadataDto {
        var dto = MetadataDto()
        val movie = movieRepo.findOne(id)
        if (movie != null) {
            dto.movieId = id
            dto.title = movie.title
            dto.year = movie.year
            dto.movieVersion = movie.version
            dto.movieCreatedDate = movie.createdDate.toString()

            val director = movie.director
            if (director != null) {
                dto.directorId = director.id
                dto.directorName = director.name
                if (director.movies != null) {
                    dto.directorOtherMovies =
                            director.movies!!
                                .filter { m -> !m.title.equals(movie.title, ignoreCase = true) }
                                .map { m -> m.title }
                                .requireNoNulls()
                }
                dto.directorVersion = director.version
                dto.directorCreatedDate = director.createdDate.toString()
            }
        }
        return dto
    }
}